google/cloud/bigtable/mutations.py

```diff
@@ -13,41 +13,200 @@
 
 
-@dataclass
 class SetCell(Mutation):
-    family: str
-    column_qualifier: bytes
-    new_value: bytes | str | int
-    timestamp_ms: int | None = None
+    def __init__(
+        self,
+        family: str,
+        qualifier: bytes | str,
+        new_value: bytes | str | int,
+        timestamp_micros: int | None = None,
+    ):
+        """
+        Mutation to set the value of a cell
+
+        Args:
+          - family: The name of the column family to which the new cell belongs.
+          - qualifier: The column qualifier of the new cell.
+          - new_value: The value of the new cell. str or int input will be converted to bytes
+          - timestamp_micros: The timestamp of the new cell. If None, the current timestamp will be used.
+              Timestamps will be sent with milisecond-percision. Extra precision will be truncated.
+              If -1, the server will assign a timestamp. Note that SetCell mutations with server-side
+              timestamps are non-idempotent operations and will not be retried.
+        """
+        qualifier = qualifier.encode() if isinstance(qualifier, str) else qualifier
+        if not isinstance(qualifier, bytes):
+            raise TypeError("qualifier must be bytes or str")
+        if isinstance(new_value, str):
+            new_value = new_value.encode()
+        elif isinstance(new_value, int):
+            new_value = new_value.to_bytes(8, "big", signed=True)
+        if not isinstance(new_value, bytes):
+            raise TypeError("new_value must be bytes, str, or int")
+        if timestamp_micros is None:
+            # use current timestamp, with milisecond precision
+            timestamp_micros = time.time_ns() // 1000
+            timestamp_micros = timestamp_micros - (timestamp_micros % 1000)
+        if timestamp_micros < SERVER_SIDE_TIMESTAMP:
+            raise ValueError(
+                "timestamp_micros must be positive (or -1 for server-side timestamp)"
+            )
+        self.family = family
+        self.qualifier = qualifier
+        self.new_value = new_value
+        self.timestamp_micros = timestamp_micros
+
+    def _to_dict(self) -> dict[str, Any]:
+        """Convert the mutation to a dictionary representation"""
+        return {
+            "set_cell": {
+                "family_name": self.family,
+                "column_qualifier": self.qualifier,
+                "timestamp_micros": self.timestamp_micros,
+                "value": self.new_value,
+            }
+        }
+
+    def is_idempotent(self) -> bool:
+        """Check if the mutation is idempotent"""
+        return self.timestamp_micros != SERVER_SIDE_TIMESTAMP
 
 
 @dataclass
 class DeleteRangeFromColumn(Mutation):
     family: str
-    column_qualifier: bytes
-    start_timestamp_ms: int
-    end_timestamp_ms: int
+    qualifier: bytes
+    # None represents 0
+    start_timestamp_micros: int | None = None
+    # None represents infinity
+    end_timestamp_micros: int | None = None
```

Comment:
I think timestamps should be longs?

Reply:
Python doesn't have a native long type. Native ints are more abstract, and have arbitrary precision: https://peps.python.org/pep-0237/

Reply:
I'm still stuck on python 2 :(